sqlite/sqlite.go

```diff
@@ -74,6 +104,163 @@ func (s *SqlStore) Flush(ctx context.Context) {
+//
+// The backup works by copying the SOURCE database to the DESTINATION database.
+// The SOURCE is the running database that needs to be backed up, and the DESTINATION
+// is the resulting backup. The underlying sqlite connection is needed for both
+// SOURCE and DESTINATION databases to use the sqlite backup API made available by the
+// go-sqlite3 driver.
+func (s *SqlStore) BackupSqlStore(ctx context.Context, w io.Writer) error {
+	span, _ := tracing.StartSpanFromContext(ctx)
+	defer span.Finish()
+
+	// create a destination db in a temporary directory to hold the backup.
+	tempDir, err := ioutil.TempDir("", "")
+	if err != nil {
+		return err
+	}
+	defer os.RemoveAll(tempDir)
+
+	destPath := tempDir + "/" + DefaultFilename
+	dest, err := NewSqlStore(destPath, zap.NewNop())
+	if err != nil {
+		return err
+	}
+
+	if err := backup(ctx, dest, s); err != nil {
+		return err
+	}
+
+	// open the backup file so it can be copied to the destination writer
+	f, err := os.Open(destPath)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	// copy the backup to the destination writer
+	_, err = io.Copy(w, f)
+	return err
+}
+
+func backup(ctx context.Context, dest, src *SqlStore) error {
+	// get the connection for the destination so we can get the underlying sqlite connection
+	destConn, err := dest.DB.Conn(ctx)
+	if err != nil {
+		return err
+	}
+	defer destConn.Close()
+
+	// get the sqlite connection for the destination to access the sqlite backup API
+	destSqliteConn, err := sqliteFromSqlConn(destConn)
+	if err != nil {
+		return err
+	}
+
+	// get the connection for the source database so we can get the underlying sqlite connection
+	srcConn, err := src.DB.Conn(ctx)
+	if err != nil {
+		return err
+	}
+	defer srcConn.Close()
+
+	// get the sqlite connection for the source to access the sqlite backup API
+	srcSqliteConn, err := sqliteFromSqlConn(srcConn)
+	if err != nil {
+		return err
+	}
+
+	// call Backup on the destination sqlite connection - which initializes the backup
+	bk, err := destSqliteConn.Backup("main", srcSqliteConn, "main")
+	if err != nil {
+		return err
+	}
+
+	// perform the backup
+	_, err = bk.Step(-1)
+	if err != nil {
+		return err
+	}
+
+	// close the backup once it's done
+	return bk.Finish()
+}
+
+// sqliteFromSqlConn returns the underlying sqlite3 connection from an sql connection
+func sqliteFromSqlConn(c *sql.Conn) (*sqlite3.SQLiteConn, error) {
+	var sqliteConn *sqlite3.SQLiteConn
+	err := c.Raw(func(driverConn interface{}) error {
+		sqliteConn = driverConn.(*sqlite3.SQLiteConn)
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return sqliteConn, nil
+}
+
+// RestoreSqlStore replaces the underlying database with the data from r.
+func (s *SqlStore) RestoreSqlStore(ctx context.Context, r io.Reader) error {
+	tempDir, err := ioutil.TempDir("", "")
+	if err != nil {
+		return err
+	}
+	defer os.RemoveAll(tempDir)
+
+	tempFileName := fmt.Sprintf("%s/%s", tempDir, DefaultFilename)
```

Comment:
Same here, could we `filepath.Join`?

Reply:
Made this change as well. Also found it funny how I did the exact same thing two different ways in the same file 😆 - now it's consistent if nothing else.